test: msg_mempool, fRelay, and other bloomfilter tests

glozow opened the pull request:
This PR adds a few tests that are bloomfilter-related, including behavior for when bloomfilters are turned _off_:
1. Tests p2p message `msg_mempool`: a node that has `peerbloomfilters` enabled should send its mempool (disabled behavior already tested [here](https://github.com/bitcoin/bitcoin/blob/master/test/functional/p2p_mempool.py)).
2. Tests that bloomfilter peers with [`fRelay=False`](https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki#extensions-to-existing-messages) in the `version` message should not receive any invs until they set the filter. The rest is the same as what’s already tested in `p2p_filter.py`.
3. Tests that peers get disconnected if they send `filterload` or `filteradd` p2p messages to a node with bloom filters disabled.
4. Refactor: renames p2p_mempool.py to p2p_nobloomfilter_messages.py.
5. Fixes race conditions in p2p_filter.py

glozow committed:
[test] add mempool msg test for node with bloomfilter enabled

-msg_mempool is currently only tested with bloomfilter disabled
(node is disconnected) in p2p_mempool.py
-msg_mempool should get mempool txns in response when bloomfilter
is enabled
-edit test that doesn't test msg_mempool as intended

glozow committed:
[test] add BIP 37 test for node with fRelay=false

A node with fRelay=False should not receive any invs until filter is set using
filterload; all other behavior should be identical.

glozow committed:
[test] sending invalid msgs to node with bloomfilters=0 causes disconnect

-A node with bloomfilters disabled should disconnect peers that send
msg_mempool, msg_filterload, or msg_filteradd.
-Renamed the test because it now has a wider scope and msg_mempool's
actual functionality makes more sense for p2p_filter.py.

jnewbery commented on test/functional/p2p_filter.py:
```diff
@@ -51,6 +56,21 @@ def on_merkleblock(self, message):
         self.tx_received = True
 
+class FilterNodeFRelay(FilterNode):
+    def peer_connect(self, *args, services=NODE_NETWORK|NODE_WITNESS, send_version=True, **kwargs):
```
I think you can remove the `services` and `send_version` arguments, and just set them directly below (this function is only ever called once, and you know what `services` and `send_version` should be set to.

jnewbery reviewed (commented):
Concept ACK. I think you can probably merge all the tests for bloom filters enabled and disabled into a single test file.

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -51,6 +56,21 @@ def on_merkleblock(self, message):
+            vt.addrFrom.port = 0
+            vt.nRelay = 0 # default = 1
+            self.on_connection_send_msg = vt  # Will be sent soon after connection_made
+        return create_conn
```
Is there any reason to overwrite this at all? The whole overwrite with the `on_connection_send_msg` magic is a bit of a mental load. I believe this can simply be a `FilterNode`, where you set `send_version=False` and then `send_message(version_without_nRelay)` and then `wait_for_verack()`.

Overall it should be less complex and less lines of code

gzhao408 replied:
Oops - had a big hole in my knowledge. How does this look:
```py
class FilterNodeWithoutNRelay(FilterNode):
    def __init__(self):
        super().__init__()
        version_without_nRelay = msg_version()
        version_without_nRelay.nRelay = 0
        self.on_connection_send_msg = version_without_nRelay
```
I'm still not 100% clear on how connection works under the hood, but setting `self.on_connection_send_msg` feels correct since it's handled in the `connection_made` callback - otherwise I need to somehow wait for connection to be established before using `send_message` directly.

MarcoFalke replied:
When I added `on_connection_send_msg` it was meant to be a "private" member. Obviously in python there is no "private", so discount my point by that.

I think either is fine. Your solution with the `on_connection_send_msg` or something along the lines of:

```
p2p = create_connection(bla)
p2p.wait_until(p2p.is_connected)
p2p.send_message(ver)
p2p.wait_for_verack()

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -51,7 +61,6 @@ def on_merkleblock(self, message):
     def on_tx(self, message):
         self.tx_received = True
 
-
```
style-nit (feel free to ignore): I think pep-8 says two new lines before `class`

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -34,6 +37,13 @@ class FilterNode(P2PInterface):
+        if fRelay == False:
+            version_without_nRelay = msg_version()
+            version_without_nRelay.nRelay = 0
+            self.on_connection_send_msg = version_without_nRelay
```
How does this work? Isn't `peer_connect` called *after* init, in which case `on_connection_send_msg` would be overwritten?

gzhao408 replied:
Aw shoot, just fixed using `send_version=False` and then manually sending it after connection is established. Asserting `relaytxes=False` so it should be set correctly now.

maflcko reviewed (commented):
Concept ACK

jnewbery commented on test/functional/p2p_filter.py:
```diff
@@ -153,6 +184,33 @@ def run_test(self):
         filter_node.send_and_ping(msg_filteradd(data=b'letstrytocrashthisnode'))
 
+    def run_test(self):
+        filter_node = self.nodes[0].add_p2p_connection(FilterNode())
```
nit: one style point that I've tried to move the tests away from is referring to connections as "nodes". A node is a bitcoind instance that we're testing. A p2p connection is a test structure used to interact with the node. Using the word _node_ for both the node and connections to it is confusing.

gzhao408 replied:
Noted! Renaming all of them increases the number of lines in the diff by a lot, so I put renames for existing code in a separate commit.

jnewbery commented on test/functional/p2p_filter.py:
```diff
@@ -93,11 +96,36 @@ def test_size_limits(self, filter_node):
+        self.nodes[0].generate(1)
 
+    def test_filter(self, filter_node):
         self.log.info('Add filtered P2P connection to the node')
```
This log seems in the wrong place. The connection has already been created earlier.

jnewbery reviewed (commented):
Looks good. A few comments inline.

gzhao408 commented on test/functional/p2p_filter.py:
```diff
@@ -125,8 +153,11 @@ def run_test(self):
 
         self.log.info('Check that we receive a tx in reply to a mempool msg if the filter matches a mempool tx')
         filter_node.merkleblock_received = False
+        filter_node.tx_received = False
```
True, I had considered just changing the log message (since it didn't seem entirely correct) but I wasn't really sure what it should be. It looks like @MarcoFalke wrote this test, I'm wondering what he would have in mind?

gzhao408 commented on test/functional/p2p_filter.py:
```diff
@@ -125,8 +153,11 @@ def run_test(self):
 
         self.log.info('Check that we receive a tx in reply to a mempool msg if the filter matches a mempool tx')
         filter_node.merkleblock_received = False
+        filter_node.tx_received = False
```
Good point. I'm thinking a getter and a setter: `get_tx_received`, `reset_tx_received` (to False), and for `merkleblock_received` as well? Shall I change the rest of the test too since it uses these throughout?

jnewbery replied:
> Shall I change the rest of the test too since it uses these throughout?

If you want to, that'd be great! There are lots of data races in this test where the test logic thread and the p2p handler thread can be accessing the same memory.

Take a look at Python properties: https://docs.python.org/3/library/functions.html#property. By implementing a getter/setter using properties, you don't need to change the client code from `assert not filter_node.tx_received` to `assert not filter_node.get_tx_received()`.

Be careful not to use the setter for the `on_xxxx()`, methods, which already have the lock when called. It's better to make the locks non-reentrant if possible (see https://github.com/bitcoin/bitcoin/pull/19178)

glozow committed:
[test] fix race conditions and test in p2p_filter

-grab mininode_lock for every access to mininode attributes,
otherwise there are race conditions

gzhao408 commented on test/functional/p2p_filter.py:
```diff
@@ -64,95 +92,140 @@ def set_test_params(self):
+        filter_peer.merkleblock_received = False
+        filter_peer.tx_received = False
         self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 90)
-        filter_node.sync_with_ping()
```
Weird that there are two `sync_with_ping`s here. @MarcoFalke would it be correct for me to remove one (since I'm touching the lines anyway)? From what I understand, one `sync_with_ping` is enough to ensure that the messages were sent and processed, and we'll be grabbing `mininode_lock` when we check the `*_received` values.

MarcoFalke replied:
There are no messages sent at all, so using `sync_with_ping` is odd, but I didn't know what else to use.

* The first sync_with_ping is needed for any imaginary invs on the wire (from `SendMessages`) to be flushed to the filter peer
* The seconds one is needed for our imaginary getdata (in reply to the imaginary inv) to be flushed to the node (and to be processed)

gzhao408 replied:
Ohh I see what you mean. If only 1 is sent, the order could be:

1. send `ping`
2. receive `inv`, send `getdata`
3. receive `pong`, can move on
5. assert not received
6. receive data (too late)

With 2, it has to be:
1. send `ping1`
2. receive `inv`, send `getdata`
3. receive `pong1`, send `ping2`
4. any response to `getdata` must be received before `pong2`
5. assert

Is this correct?

MarcoFalke replied:
Jup, that's the idea

MarcoFalke commented:
nit: commit f78730d582c605882e5a3bf554aefb1a92942b0e can be done as a scripted diff, but no big deal

glozow committed:
scripted-diff: rename node to peer for mininodes

-BEGIN VERIFY SCRIPT-
sed -i 's/FilterNode/P2PBloomFilter/g' test/functional/p2p_filter.py;
sed -i 's/filter_node/filter_peer/g' test/functional/p2p_filter.py;
-END VERIFY SCRIPT-

gzhao408 commented:
Resolves #18473. Ready for Review again!

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -93,6 +94,23 @@ def test_size_limits(self, filter_node):
+        self.log.info("Check that a node with bloom filters enabled services p2p mempool messages")
+        self.nodes[0].disconnect_p2ps()
+        filter_peer = FilterNode()
+        filter_peer.tx_received = False
```
in commit 75865b28a0:

```suggestion
```

I think tx_received  can be removed from this test, as it is redundant to `wait_for_tx`, no?

gzhao408 replied:
Er, do you mean because it's already set to False in the constructor? I removed it for this reason, thanks for bringing it to my attention.
I don't think we can remove `tx_received` because we sometimes want to assert that we didn't receive a tx with it.

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -125,8 +143,11 @@ def run_test(self):
 
         self.log.info('Check that we receive a tx in reply to a mempool msg if the filter matches a mempool tx')
```
same commit:

This comment is obviously wrong (`git blame myself`)

```suggestion
        self.log.info('Check that we receive a tx if the filter matches a mempool tx')
```

MarcoFalke replied:
unrelated nit in commit facd5f9765:

Has this comment been fixed?

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -125,8 +143,11 @@ def run_test(self):
         filter_node.merkleblock_received = False
+        filter_node.tx_received = False
         txid = self.nodes[0].sendtoaddress(filter_address, 90)
+        self.nodes[0].p2p.send_message(msg_mempool())
```
So this change needs to go

gzhao408 replied:
I removed it altogether, since receiving tx behavior for this reason is already tested and `msg_mempool` is added.

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -93,6 +94,23 @@ def test_size_limits(self, filter_node):
 
+    def test_msg_mempool(self):
+        self.log.info("Check that a node with bloom filters enabled services p2p mempool messages")
+        self.nodes[0].disconnect_p2ps()
```
Same commit:

I have a slight preference that tests clean up after themselves, and not after other tests. So this should be moved to the end of the subtest that created the peer connection.

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -174,6 +184,32 @@ def run_test(self):
+        self.log.info('Test BIP 37 for a node with fRelay = True (default)')
+        self.test_filter(filter_peer)
+        self.nodes[0].disconnect_p2ps()
+        wait_until(lambda: len(self.nodes[0].getpeerinfo()) == 0)
```
in commit 5f0dd129a0:

It seems a burden on tests to call this to avoid races. It might be easy to forget and races might only appear long after the test has been written and merged. It seems less fragile to move this wait into the disconnect helper.

gzhao408 replied:
Uh oh... I naively added a `wait_until(lambda: len(self.getpeerinfo()) == 0)` but `getpeerinfo` gets both p2p connections and connected nodes. I need to figure out a better way to do this.

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -174,6 +184,32 @@ def run_test(self):
+        self.test_frelay_false(filter_peer_without_nrelay)
+        self.test_filter(filter_peer_without_nrelay)
+        self.nodes[0].disconnect_p2ps()
+        wait_until(lambda: len(self.nodes[0].getpeerinfo()) == 0)
```
same

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -174,6 +184,32 @@ def run_test(self):
+        filter_peer_without_nrelay.wait_for_verack()
+        assert not self.nodes[0].getpeerinfo()[0]['relaytxes']
+        self.test_frelay_false(filter_peer_without_nrelay)
+        self.test_filter(filter_peer_without_nrelay)
```
Same commit: Could add `assert self.nodes[0].getpeerinfo()[0]['relaytxes']` after this line?

gzhao408 replied:
Added it right after it should be set to True after the `filterload` message in `test_filter`.

maflcko reviewed (commented):
ACK 1e64440490 looks like an improvement, but I have some nits to consider 🍬

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 1e64440490 looks like an improvement, but I have some nits to consider 🍬
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiqqQv9HPT78+tE2ecU5neOV9/iItDomZZ3+dJ/+1clX3pZNzJ7k0Qh/Cr3DBU+
Eh4zIS0D1ooVyxknFMhGziukNAmlzYAFW1lSytCwEMCxZhJ4Nujlq3dp/aaQ5FoH
j1D2zvTe4QEKvysqTqHafJQ0iuxYoXzO2TXs5io6923GpxLHwndOiozkNUkLd9TA
2W2XwhO/k7Uu0r0SCBFmSyZmb93Jus8h118q0Dskg7uM8FBJ/FuPS6xvIUWpn21h
3etYU0+9E8ncRJ6SXTU4bKcPsHbJrpjo3hzGghhJLlz1+ywJzSiF2TVumFilgMyU
+MwS/Yvm+vwuTQwFOzIzAQfQYdm1Dh0JOcy52Uob1keytZV/K3opDS3MrlvlPxIC
4ZKMuT/JclX9UkKzIu/ppQ6O57PbhJAcZyXBs4JrQ896rLccsYSGO3K1jQgijrEA
lZQhSVPm2Y0/cHMrgr9/QcZKk+Ffyvz89HgThyrvmNJNdGCLhuPnoWlQMwjhpviq
akjJhYf+
=Pmtq
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `88b89b80c24192ef59fea53c027a40ed6e6fe7c8d6795036ca7991f3fb4d7971  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010888b89b80c24192ef59fea53c027a40ed6e6fe7c8d6795036ca7991f3fb4d7971f0100ac81399dec8cb2b0557c096e7bb137708fff01081e2a4e51dab34de2da88ceea0f44e1008f1045edf8643f00884179ff50fbd12040083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010ac99af0bebc93ec238fdb910a9d9404c08f1045edf8643f008c7feb5b30f1d95650083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010e473312931eefdec8ffbdba7e1c5164d08f1045edf8643f008ea77f5def83620d00083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010f51726529a632013f4bec2a1743d00d808f1045edf8643f0081b99dd895e2e9f630083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -93,6 +94,21 @@ def test_size_limits(self, filter_node):
+        filter_peer.send_and_ping(filter_peer.watch_filter_init)
+        self.nodes[0].p2p.send_message(msg_mempool())
+        filter_peer.wait_for_tx(txid)
+        assert filter_peer.tx_received
```
nit in commit facd5f9765:

This assert is not needed. A stronger assert was checked in the previous line, no?

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -123,12 +139,6 @@ def run_test(self):
-        filter_node.merkleblock_received = False
-        txid = self.nodes[0].sendtoaddress(filter_address, 90)
-        filter_node.wait_for_tx(txid)
-        assert not filter_node.merkleblock_received
```
facd5f9765 why is this test removed?

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -109,13 +110,23 @@ def test_msg_mempool(self):
+    def test_filter(self, filter_node, frelay=False):
+        # Set the bloomfilter using filterload
         filter_node.send_and_ping(filter_node.watch_filter_init)
+        if frelay == False:
```
in 27f3cac34e

why is this guard needed?

maflcko reviewed (commented):
Concept ACK b758749173

jonatack commented on test/functional/p2p_filter.py:
```diff
@@ -64,95 +91,144 @@ def set_test_params(self):
+        filter_peer = P2PBloomFilter()
 
-        self.test_size_limits(filter_node)
+        self.log.info("Create a tx relevant to the peer before connecting")
```
e8acc60 It looks like these logs lines 127 and 131 are related to the same test as the logging at line 124. Perhaps make them `debug` logs instead of `info` to declutter the ample logging output. (Thanks for putting the test in a separate function as you did.)

jonatack commented on test/functional/p2p_filter.py:
497a6193 Is this logging (or the one at the start of `test_msg_mempool()`) redundant?

jonatack commented on test/functional/p2p_filter.py:
497a6193 perhaps put the BIP37 tests into their own function

gzhao408 replied:
They're mostly in their own function `test_filter`, so I left this one as is.

jonatack commented on test/functional/p2p_nobloomfilter_messages.py:
```diff
@@ -0,0 +1,43 @@
+from test_framework.util import assert_equal
+
+
+class P2PNobloomfilterMessages(BitcoinTestFramework):
```
4ef80f0 nit: `P2PNoBloomFilterMessages` (and at the last line)

jonatack commented on test/functional/p2p_nobloomfilter_messages.py:
```diff
@@ -0,0 +1,43 @@
+        self.extra_args = [["-peerbloomfilters=0"]]
+
+    def test_message_causes_disconnect(self, message):
+        # Add a p2p connection that sends a message and check that it disconnects
```
4ef80f0 nit: this could be a `"""docstring."""`

jonatack commented on test/functional/p2p_nobloomfilter_messages.py:
```diff
@@ -0,0 +1,43 @@
+
+        self.log.info("Test that node is disconnected if it sends filteradd message")
+        self.test_message_causes_disconnect(msg_filteradd(data=b'\xcc'))
+
```
4ef80f0 nit: I think PEP8 (or this codebase, not sure which) prefers two lines here. Same in p2p_filter.py in dca73941.

jonatack reviewed (commented):
ACK dca73941eb0f0a4c9b68efed3870b536f7dd6cfe modulo a few nits if you retouch, happy to re-ACK if you take any of them but don't feel obliged to.

MarcoFalke commented:
With two ACKs, this test change seems ready to merge. @gzhao408 Let me know if you want to address the feedback or want to leave it as is for now (for potential follow-ups) and have the pull request merged.

gzhao408 commented:
Thank you for the review @jonatack 🙏 ! 
I've made a followup #19252 for `disconnect_p2ps` to wait for disconnect to complete anyway, so I can easily incorporate the feedback there. @MarcoFalke green light to merge :) thanks
